File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/network/PaymentByBalanceRequest.java
package hottopic.mit.co.nz.cleaningservice.network;

import android.content.Context;

import java.util.Map;

import hottopic.mit.co.nz.cleaningservice.entities.network.params.PaymentParams;
import hottopic.mit.co.nz.cleaningservice.entities.network.response.UserInfoResponse;
import hottopic.mit.co.nz.cleaningservice.network.service.PaymentService;
import io.reactivex.Single;

public class PaymentByBalanceRequest extends BaseRequest<PaymentService, UserInfoResponse, UserInfoResponse> {
    private PaymentParams paymentParams;

    public PaymentByBalanceRequest(Context context, PaymentParams paymentParams) {
        super(context);
        this.paymentParams = paymentParams;
    }

    @Override
    protected Map<String, String> getHeader() {
        return null;
    }

    @Override
    protected Class<PaymentService> getEndpointClass() {
        return PaymentService.class;
    }

    @Override
    protected Single<UserInfoResponse> getEndpoint(PaymentService endpoint) {
        return endpoint.paymentByBalance(paymentParams);
    }

    @Override
    protected UserInfoResponse dealWithResponse(UserInfoResponse response) {
        return response;
    }
}

File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/network/PaymentByBalanceRequest.kt
package hottopic.mit.co.nz.cleaningservice.network

import android.content.Context

import hottopic.mit.co.nz.cleaningservice.entities.network.params.PaymentParams
import hottopic.mit.co.nz.cleaningservice.entities.network.response.UserInfoResponse
import hottopic.mit.co.nz.cleaningservice.network.service.PaymentService
import io.reactivex.Single

class PaymentByBalanceRequest(context: Context, private val paymentParams: PaymentParams) : BaseRequest<PaymentService, UserInfoResponse, UserInfoResponse>(context) {

    override val header: Map<String, String>?
        get() = null

    override val endpointClass: Class<PaymentService>
        get() = PaymentService::class.java

    override fun getEndpoint(endpoint: PaymentService?): Single<UserInfoResponse?>? {
        return endpoint?.paymentByBalance(paymentParams)
    }

    override fun dealWithResponse(response: UserInfoResponse): UserInfoResponse? {
        return response
    }
}
